### PR TITLE
Optimize Recommendation Validator

### PR DESCRIPTION
**Pull request by abs51295**
Currently, post filter takes almost 7.5 milliseconds to validate the recommendations obtained from PGM. Reducing the post filter time to 0.75 milliseconds.

### CODE REVIEW

**Comment by rootAvish** [comment]
[test]

File: analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py [hunk 1]
```diff
@@ -76,7 +76,7 @@ def _generate_cooccurrence_matrix_for_ecosystem(cls, list_of_package_list, node_
         row_count = len(list_of_package_list)
         cooccurrence_matrix = utils.create_empty_pandas_df(
             rowsize=row_count, columns_list=node_list)
-        component_class_list = kronos_component_dependency_dict.keys()
+        component_class_list = list(kronos_component_dependency_dict)
```
Comment (rootAvish) [inline analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py]: Use .keys() here.

File: analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py [hunk 2]
```diff
@@ -98,7 +98,7 @@ def _generate_cooccurrence_matrix_for_ecosystem(cls, list_of_package_list, node_
                     [row_id], component_class] = component_class_occurrence
 
             temp_node_list = list(component_class_list)
-            intent_list = kronos_intent_dependency_dict.keys()
+            intent_list = list(kronos_intent_dependency_dict)
```
Comment (rootAvish) [inline analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py]: Here as well.

File: analytics_platform/kronos/src/recommendation_validator.py [hunk 3]
```diff
@@ -23,24 +23,25 @@ def load_package_list(input_manifest_data_store, additional_path, input_ecosyste
         :param additional_path: The directory to pick the manifest files from.
         :param input_ecosystem: The ecosystem for which the aggregated manifest list will be saved.
 
-        :return: RecommendationValidator object."""
+        :return: RecommendationValidator object.
 
-        all_list_of_package_list = []
+        NOTE: This method is called only once while bringing the kronos api up."""
```
Comment (rootAvish) [inline analytics_platform/kronos/src/recommendation_validator.py]: This note goes before the params.

File: analytics_platform/kronos/src/recommendation_validator.py [hunk 4]
```diff
@@ -23,24 +23,25 @@ def load_package_list(input_manifest_data_store, additional_path, input_ecosyste
...
         manifest_filenames = input_manifest_data_store.list_files(
             additional_path + MANIFEST_FILEPATH)
         for manifest_filename in manifest_filenames:
-            user_category = manifest_filename.split("/")[-2]
             manifest_content_json_list = input_manifest_data_store.read_json_file(
                 filename=manifest_filename)
             for manifest_content_json in manifest_content_json_list:
                 manifest_content_dict = dict(manifest_content_json)
-                ecosystem = manifest_content_dict.get(MANIFEST_ECOSYSTEM)
+                ecosystem = manifest_content_dict[MANIFEST_ECOSYSTEM]
```
Comment (rootAvish) [inline analytics_platform/kronos/src/recommendation_validator.py]: Why are we not using `.get` here? Are we sure this key will always exist in the dict?

File: analytics_platform/kronos/src/recommendation_validator.py [hunk 5]
```diff
@@ -23,24 +23,25 @@ def load_package_list(input_manifest_data_store, additional_path, input_ecosyste
...
                 if ecosystem != input_ecosystem:
                     continue
-                list_of_package_list = manifest_content_dict.get(
-                    MANIFEST_PACKAGE_LIST)
-                all_list_of_package_list.extend(list_of_package_list)
-        return RecommendationValidator(all_list_of_package_list)
+                list_of_package_list = manifest_content_dict[MANIFEST_PACKAGE_LIST]
```
Comment (rootAvish) [inline analytics_platform/kronos/src/recommendation_validator.py]: Same here. Why remove `.get`?

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 4 above)
Comment (sara-02) [reply analytics_platform/kronos/src/recommendation_validator.py]: And return as `.get(MANIFEST_ECOSYSTEM, "")`

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 5 above)
Comment (sara-02) [reply analytics_platform/kronos/src/recommendation_validator.py]: return as `.get(MANIFEST_PACKAGE_LIST, [])`

File: analytics_platform/kronos/src/recommendation_validator.py [hunk 6]
```diff
@@ -82,9 +83,9 @@ def generate_companion_dependency_set(self, input_list, companion_package):
         :return: A set() of user_stack + companion_package."""
 
         if input_list is not None and companion_package is not None:
-            input_list = input_list[:]
-            input_list.append(companion_package)
-            return set(input_list)
+            input_set = set(input_list)
+            input_set.add(companion_package)
+            return input_set
```
Comment (rootAvish) [inline analytics_platform/kronos/src/recommendation_validator.py]: 86-88 can be compressed into 

```python 
return set(input_list).union([companion_package])
```

File: analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py (hunk 1 above)
Comment (abs51295) [reply analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py]: Sure

File: analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py (hunk 2 above)
Comment (abs51295) [reply analytics_platform/kronos/softnet/src/cooccurrence_matrix_generator.py]: Sure

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 3 above)
Comment (abs51295) [reply analytics_platform/kronos/src/recommendation_validator.py]: Moving it before params.

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 4 above)
Comment (abs51295) [reply analytics_platform/kronos/src/recommendation_validator.py]: Thanks @sara-02 for the suggestion :+1:

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 5 above)
Comment (abs51295) [reply analytics_platform/kronos/src/recommendation_validator.py]: Following @sara-02's suggestion.

File: analytics_platform/kronos/src/recommendation_validator.py (hunk 6 above)
Comment (abs51295) [reply analytics_platform/kronos/src/recommendation_validator.py]: Nice suggestion :+1: Thanks.

File: analytics_platform/kronos/src/recommendation_validator.py [hunk 7]
```diff
@@ -97,51 +98,27 @@ def generate_alternate_dependency_set(self, input_list, alternate_package, alter
...
-        :return: Frequency of subset where test stack is a part of any known dependency list."""
+        generate_companion_dependency_set() or generate_alternate_dependency_set().
 
+        :return: Frequency of subset where test stack is a part of any known dependency set."""
         count = 0
-        for dependency_list in self.all_list_of_package_list:
-            manifest_dependency_set = set(dependency_list)
-            if manifest_dependency_set >= recommended_dependency_set:
+        for dependency_set in self.all_list_of_package_set:
+            if dependency_set >= recommended_dependency_set:
```
Comment (rootAvish) [inline analytics_platform/kronos/src/recommendation_validator.py]: Use `.issubset` method here.

Comment (abs51295) [reply analytics_platform/kronos/src/recommendation_validator.py]: Sure

File: analytics_platform/kronos/src/kronos_online_scoring.py [hunk 8]
```diff
@@ -263,7 +263,7 @@ def get_alternate_packages_for_packages(similar_package_dict, package_names,
         if num_alternate_packages > alt_package_count_threshold:
             num_alternate_packages = alt_package_count_threshold
         alternate_package_dict_list_of_package_pruned = alternate_package_dict_list_of_package[
-            :num_alternate_packages]
+                                                        :num_alternate_packages]
```
Comment (rootAvish) [inline analytics_platform/kronos/src/kronos_online_scoring.py]: Fix the indent here.

File: tests/unit_tests/test_recommendation_validator.py [hunk 9]
```diff
@@ -90,7 +89,7 @@ def test_generate_companion_set(self):
         companion_package = 'D'
         comp_set = obj.generate_companion_dependency_set(
             input_list, companion_package)
-        self.assertTrue(comp_set == set(['A', 'B', 'C', 'D']))
+        self.assertTrue(comp_set == {'A', 'B', 'C', 'D'})
```
Comment (rootAvish) [inline tests/unit_tests/test_recommendation_validator.py]: `assertEqual` here.

File: tests/unit_tests/test_recommendation_validator.py [hunk 10]
```diff
@@ -100,35 +99,29 @@ def test_generate_alternate_set(self):
         alternate_to = 'B'
         alt_set = obj.generate_alternate_dependency_set(
             input_list, alternate_package, alternate_to)
-        self.assertTrue(alt_set == set(['A', 'B123', 'C']))
+        self.assertTrue(alt_set == {'A', 'B123', 'C'})
```
Comment (rootAvish) [inline tests/unit_tests/test_recommendation_validator.py]: Here as well.

File: tests/unit_tests/test_recommendation_validator.py [hunk 11]
```diff
@@ -100,35 +99,29 @@ def test_generate_alternate_set(self):
...
-            recommended_dependency_set, input_list)
-        self.assertTrue(len(diff_list) == 3)
-        self.assertTrue(diff_list == ['E', 'E', 'F'])
-        recommended_dependency_set = set(['A', 'B', 'C', 'D', 'Q'])
-        diff_list = obj.check_companion_recommendation_validity(
-            recommended_dependency_set, input_list)
-        self.assertTrue(len(diff_list) == 0)
+        recommended_dependency_set = {'A', 'B', 'C', 'D', 'E'}
+        count = obj.check_companion_or_alternate_recommendation_validity(recommended_dependency_set)
+        self.assertTrue(count == 2)
```
Comment (rootAvish) [inline tests/unit_tests/test_recommendation_validator.py]: Same.

File: tests/unit_tests/test_recommendation_validator.py [hunk 12]
```diff
@@ -100,35 +99,29 @@ def test_generate_alternate_set(self):
...
+        recommended_dependency_set = {'A', 'B', 'C', 'D', 'Q'}
+        count = obj.check_companion_or_alternate_recommendation_validity(recommended_dependency_set)
+        self.assertTrue(count == 0)
```
Comment (rootAvish) [inline tests/unit_tests/test_recommendation_validator.py]: Use `assertEqual` here.

File: tests/unit_tests/test_recommendation_validator.py [hunk 13]
```diff
@@ -182,4 +175,4 @@ def test_check_companion_recommendation(self):
             input_list, companion_packages)
         self.assertTrue(len(result_companion_list) == 2)
         comp_name = [c.get('package_name') for c in result_companion_list]
-        self.assertTrue(set(comp_name) == set(['E', 'F']))
+        self.assertTrue(set(comp_name) == {'E', 'F'})
```
Comment (rootAvish) [inline tests/unit_tests/test_recommendation_validator.py]: Same.

**Pull request merged**